libsawtooth/src/client/mod.rs

```diff
@@ -0,0 +1,155 @@
+ */
+
+mod error;
+
+use reqwest::{blocking::Client, Url};
+use serde::Deserialize;
+use std::collections::VecDeque;
+
+pub use error::SawtoothClientError;
+
+/// Client for Sawtooth Cli transactions
+/// Handles communication with the REST API
+pub struct RestApiSawtoothClient {
+    url: String,
+}
+
+trait SawtoothClient {
+    fn new(url: &str) -> Self;
+    fn list_batches(&self) -> Result<Box<dyn Iterator<Item = Batch>>, SawtoothClientError>;
+}
+
+impl SawtoothClient for RestApiSawtoothClient {
+    /// Create a new `RestApiSawtoothClient` with the given base `url`. The URL should be the bind
+    /// endpoint of the Sawtooth REST API
+    fn new(url: &str) -> RestApiSawtoothClient {
+        RestApiSawtoothClient { url: url.into() }
+    }
+
+    /// Lists all existing batches
+    fn list_batches(&self) -> Result<Box<dyn Iterator<Item = Batch>>, SawtoothClientError> {
+        let url = format!("{}/batches", &self.url);
+
+        Ok(Box::new(BatchIter::new(&url).unwrap()))
+    }
+}
+
+/// Iterator used for parsing and deserializing batches
+pub struct BatchIter {
+    next: String,
+    cache: VecDeque<Batch>,
+}
+
+impl BatchIter {
+    /// Create a new 'BatchIter' which will make a call to the REST API and load the initial
+    /// cache with the first page of batch entries
+    fn new(url: &str) -> Result<Self, SawtoothClientError> {
+        let url = Url::parse(url)
+            .map_err(|err| SawtoothClientError::new_with_source("invalid URL", err.into()))?;
+
+        let request = Client::new().get(url);
+        let response = request
+            .send()
+            .map_err(|err| SawtoothClientError::new_with_source("request failed", err.into()))?;
+
+        let page: Page = response.json().map_err(|err| {
+            SawtoothClientError::new_with_source("failed to deserialize response body", err.into())
+        })?;
+
+        let init_cache = page.data.into();
+
+        let next = page.next;
+
+        Ok(Self {
+            next,
+            cache: init_cache,
+        })
+    }
+
+    /// If another page of batches exisits, use the 'next' url from the current page and
+    /// reload the cache with the next page of batch entries
+    fn reload_cache(&mut self) -> Result<(), SawtoothClientError> {
+        let request = Client::new().get(&self.next);
+        let response = request
+            .send()
+            .map_err(|err| SawtoothClientError::new_with_source("request failed", err.into()))?;
+
+        let page: Page = response.json().map_err(|err| {
+            SawtoothClientError::new_with_source("failed to deserialize response body", err.into())
+        })?;
+
+        self.cache = page.data.into();
+
+        self.next = page.next;
+
+        Ok(())
+    }
+}
+
+/// Return the next batch from the cache, if the cache is empty reload it
+impl Iterator for BatchIter {
+    type Item = Batch;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cache.is_empty() && self.next != "null" {
```

Comment:
if next is `Option<String>`, this check can be `self.next.is_some()`

Reply:
Fixed